lib/posix/fs.c

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <string.h>
+#include <errno.h>
+#include <posix/pthread.h>
+#include <posix/unistd.h>
+
+union file_desc {
+	struct fs_file_t file;
+	struct fs_dir_t	dir;
+};
+
+struct posix_fs_desc {
+	union file_desc desc;
+	bool is_dir;
+	bool used;
+};
+
+static struct posix_fs_desc desc_array[CONFIG_MAX_OPEN_FILES];
+
+static struct fs_dirent fdirent;
+static struct dirent pdirent;
+
+#ifdef CONFIG_ERRNO
+static inline void set_pthread_errno(int err)
+{
+	unsigned int key = irq_lock();
+
+	((struct k_thread *)pthread_self())->errno_var = err;
+	irq_unlock(key);
+}
+#else
+#define set_pthread_errno(x) do { } while (0)
+#endif
+
+static int posix_fs_alloc_fd(void **ptr, bool is_dir)
+{
+	int fd;
+	unsigned int key = irq_lock();
+
+	for (fd = 0; fd < CONFIG_MAX_OPEN_FILES; fd++) {
+		if (desc_array[fd].used == false) {
+			if (is_dir) {
+				*ptr = &desc_array[fd].desc.dir;
+			} else {
+				*ptr = &desc_array[fd].desc.file;
+			}
+			desc_array[fd].used = true;
+			desc_array[fd].is_dir = is_dir;
+			break;
+		}
+	}
+	irq_unlock(key);
+
+	if (fd >= CONFIG_MAX_OPEN_FILES) {
+		return -1;
+	}
+
+	return fd;
+}
+
+static void posix_fs_free_fd(int fd)
+{
+	unsigned int key = irq_lock();
+
+	desc_array[fd].used = false;
+	desc_array[fd].is_dir = false;
+	irq_unlock(key);
+}
+
+static void posix_fs_free_ptr(void *ptr)
+{
+	struct posix_fs_desc *desc = (struct posix_fs_desc *)ptr;
+	unsigned int key = irq_lock();
+
+	desc->used = false;
+	desc->is_dir = false;
+	irq_unlock(key);
+}
+
+static int posix_fs_get_ptr(int fd, void **ptr, bool is_dir)
+{
+	int rc = 0;
+	unsigned int key;
+
+	if (fd < 0 || fd >= CONFIG_MAX_OPEN_FILES) {
+		return -1;
+	}
+
+	key = irq_lock();
+
+	if ((desc_array[fd].used == true) &&
+		(desc_array[fd].is_dir == is_dir)) {
+		if (is_dir) {
+			*ptr = &desc_array[fd].desc.dir;
+		} else {
+			*ptr = &desc_array[fd].desc.file;
+		}
+	} else {
+		rc = -1;
+	}
+	irq_unlock(key);
+
+	return rc;
+}
+
+/**
+ * @brief Open a file.
+ *
+ * See IEEE 1003.1
+ */
+int open(const char *name, int flags)
+{
+	int rc, fd;
+	struct fs_file_t *ptr = NULL;
+
+	ARG_UNUSED(flags);
+
+	fd = posix_fs_alloc_fd((void **)&ptr, false);
+	if ((fd < 0) || (ptr == NULL)) {
+		set_pthread_errno(ENFILE);
+		return -1;
+	}
+	memset(ptr, 0, sizeof(struct fs_file_t));
+
+	rc = fs_open(ptr, name);
+	if (rc < 0) {
+		posix_fs_free_fd(fd);
+		set_pthread_errno(-rc);
+		return -1;
+	}
+
+	return fd;
+}
+
+/**
+ * @brief Close a file descriptor.
+ *
+ * See IEEE 1003.1
+ */
+int close(int fd)
```

Comment:
Handles are not released when threads exit.  Zephyr does not have the notion of processes, so this is most likely fine as it's the expected behavior.  What does the POSIX spec say on the subject, especially when targeting embedded devices without processes?